actionview/lib/action_view/helpers/translation_helper.rb

```diff
@@ -69,78 +69,43 @@ module TranslationHelper
       # resolved against.
       #
       def translate(key, **options)
-        unless options[:default].nil?
-          remaining_defaults = Array.wrap(options.delete(:default)).compact
-          options[:default] = remaining_defaults unless remaining_defaults.first.kind_of?(Symbol)
-        end
+        return key.map { |k| translate(k, **options) } if key.is_a?(Array)
 
-        # If the user has explicitly decided to NOT raise errors, pass that option to I18n.
-        # Otherwise, tell I18n to raise an exception, which we rescue further in this method.
-        # Note: `raise_error` refers to us re-raising the error in this method. I18n is forced to raise by default.
-        if options[:raise] == false
-          raise_error = false
-          i18n_raise = false
-        else
-          raise_error = options[:raise] || ActionView::Base.raise_on_missing_translations
-          i18n_raise = true
+        defaults = if options.key?(:default)
+          options[:default].is_a?(Array) ? options[:default].compact : [options[:default]]
         end
 
-        fully_resolved_key = scope_key_by_partial(key)
-
-        if html_safe_translation_key?(key)
-          html_safe_options = options.dup
+        options[:raise] = true if options[:raise].nil? && ActionView::Base.raise_on_missing_translations
+        guard = MISSING_TRANSLATION
 
-          options.except(*I18n::RESERVED_KEYS).each do |name, value|
-            unless name == :count && value.is_a?(Numeric)
-              html_safe_options[name] = ERB::Util.html_escape(value.to_s)
-            end
+        translation = while key
+          if defaults.blank? && !options[:raise].nil?
+            guard = [] # nullify guard, let I18n handle missing translation
```

Comment:
I'd prefer to extract a constant here, not just for performance but also to give the behaviour a name; `[]` seems like it should disable something or be a no-op in some way, when it actually enables i18n's missing translation handler.

I'd also call this `default`; the name `guard` still sets off my "something intricate and important is happening here!" sense in the same way that `sentinel` did, when we haven't even gotten to the important part of the method yet.

```patch
diff --git a/actionview/lib/action_view/helpers/translation_helper.rb b/actionview/lib/action_view/helpers/translation_helper.rb
index 2c804f0736..8122f0470e 100644
--- a/actionview/lib/action_view/helpers/translation_helper.rb
+++ b/actionview/lib/action_view/helpers/translation_helper.rb
@@ -75,11 +75,11 @@ def translate(key, **options)
         end

         options[:raise] = true if options[:raise].nil? && ActionView::Base.raise_on_missing_translations
-        guard = MISSING_TRANSLATION
+        default = MISSING_TRANSLATION

         translation = while key
           if defaults.blank? && !options[:raise].nil?
-            guard = [] # nullify guard, let I18n handle missing translation
+            default = HANDLE_WITH_I18N
           end

           key = scope_key_by_partial(key)
@@ -87,10 +87,10 @@ def translate(key, **options)

           if html_safe_translation_key?(key)
             html_safe_options ||= html_escape_translation_options(options)
-            translated = I18n.translate(key, **html_safe_options, default: guard)
+            translated = I18n.translate(key, **html_safe_options, default: default)
             break html_safe_translation(translated) unless translated.equal?(MISSING_TRANSLATION)
           else
-            translated = I18n.translate(key, **options, default: guard)
+            translated = I18n.translate(key, **options, default: default)
             break translated unless translated.equal?(MISSING_TRANSLATION)
           end

@@ -121,6 +121,9 @@ def localize(object, **options)
         MISSING_TRANSLATION = Object.new
         private_constant :MISSING_TRANSLATION

+        HANDLE_WITH_I18N = [].freeze
+        private_constant :HANDLE_WITH_I18N
+
         def self.i18n_option?(name)
           (@i18n_option_names ||= I18n::RESERVED_KEYS.to_set).include?(name)
         end
```

Feel free to disagree with me here (or in any other case!) by the way: I'm happy enough with the current code, this really is just a suggestion.

Reply:
I've renamed `guard` to `default` and `defaults` to `alternatives`.  I felt like having both `default` and `defaults` was too confusing, especially since they aren't actually connected.  Overall I think it's a big improvement! :+1:

I also extracted the constant, but it felt marginally more natural to name the value by what it is rather than how it's used, so I named it `NO_DEFAULT`.  (And I left the code comment highlighting the hand-off to I18n.)  Any objections?

<details>
  <summary>diff</summary>
  
  ```diff
  diff --git a/actionview/lib/action_view/helpers/translation_helper.rb b/actionview/lib/action_view/helpers/translation_helper.rb
  index 302dcfe6d8..513290275b 100644
  --- a/actionview/lib/action_view/helpers/translation_helper.rb
  +++ b/actionview/lib/action_view/helpers/translation_helper.rb
  @@ -70,16 +70,16 @@ module TranslationHelper
         def translate(key, **options)
           return key.map { |k| translate(k, **options) } if key.is_a?(Array)

  -        defaults = if options.key?(:default)
  +        alternatives = if options.key?(:default)
             options[:default].is_a?(Array) ? options.delete(:default).compact : [options.delete(:default)]
           end

           options[:raise] = true if options[:raise].nil? && ActionView::Base.raise_on_missing_translations
  -        guard = MISSING_TRANSLATION
  +        default = MISSING_TRANSLATION

           translation = while key
  -          if defaults.blank? && !options[:raise].nil?
  -            guard = [] # nullify guard, let I18n handle missing translation
  +          if alternatives.blank? && !options[:raise].nil?
  +            default = NO_DEFAULT # let I18n handle missing translation
             end

             key = scope_key_by_partial(key)
  @@ -87,16 +87,16 @@ def translate(key, **options)

             if html_safe_translation_key?(key)
               html_safe_options ||= html_escape_translation_options(options)
  -            translated = I18n.translate(key, **html_safe_options, default: guard)
  +            translated = I18n.translate(key, **html_safe_options, default: default)
               break html_safe_translation(translated) unless translated.equal?(MISSING_TRANSLATION)
             else
  -            translated = I18n.translate(key, **options, default: guard)
  +            translated = I18n.translate(key, **options, default: default)
               break translated unless translated.equal?(MISSING_TRANSLATION)
             end

  -          break defaults.first if defaults.present? && !defaults.first.is_a?(Symbol)
  +          break alternatives.first if alternatives.present? && !alternatives.first.is_a?(Symbol)

  -          key = defaults&.shift
  +          key = alternatives&.shift
           end

           if key.nil?
  @@ -121,6 +121,9 @@ def localize(object, **options)
           MISSING_TRANSLATION = Object.new
           private_constant :MISSING_TRANSLATION

  +        NO_DEFAULT = [].freeze
  +        private_constant :NO_DEFAULT
  +
           def self.i18n_option?(name)
             (@i18n_option_names ||= I18n::RESERVED_KEYS.to_set).include?(name)
           end
  ```
</details>

Reply:
💯 `alternatives` is great, much clearer!